File: 198.py
class Solution(object):
    def rob(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if not nums: return 0
        if len(nums) == 1: return nums[0]
        current_value, max_value = nums[0], max(nums[0], nums[1])
        for i in range(2,len(nums)):
            max_value, current_value = max(max_value, current_value + nums[i]) ,max_value
        return max_value

File: 198.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        if (nums.size() == 0){
            return 0;
        }
        if (nums.size() == 1){
            return nums[0];
        }
        int m = nums.size();
        vector<int> res(m, 0);
        res[0] = nums[0];
        res[1] = max({nums[0], nums[1]});
        for (int i =2 ; i < nums.size(); i++){
            res[i] = max(res[i-2] + nums[i], res[i-1]);
        }
        return res[m-1];
    }
};
